cirq/contrib/noise_models/noise_models.py

```diff
@@ -0,0 +1,152 @@
+    """Whether the moment is nothing but measurement gates.
+
+    If a moment is a mixture of measurement and non-measurement gates
+    this will throw a ValueError.
+    """
+    is_measurements = None
+    for gate in moment:
+        if protocols.is_measurement(gate):
+            if is_measurements is None:
+                is_measurements = True
+            elif is_measurements:
+                # Check that *all* are measurements
+                pass
+            else:
+                raise ValueError("Moment must be all measurements "
+                                 "or all operations.")
+        else:
+            if is_measurements is None:
+                is_measurements = False
+            elif not is_measurements:
+                # Check that all are *not* measurements
+                pass
+            else:
+                raise ValueError("Moment must be all measurements "
+                                 "or all operations.")
+    return is_measurements
+
+
+class DepolarizingNoiseModel(devices.NoiseModel):
+    """Applies depolarizing noise to each qubit individually at the end of
+    every moment.
+
+    If a circuit contains measurements, they be in moments that don't
+    also contain gates.
+
+    Args:
+        depol_prob: Depolarizing probability.
+    """
+
+    def __init__(self, depol_prob: float):
+        value.validate_probability(depol_prob, 'depol prob')
+        self.qubit_noise_gate = ops.DepolarizingChannel(depol_prob)
+
+    def noisy_moment(self, moment: 'cirq.Moment',
+                     system_qubits: Sequence['cirq.Qid']):
+        if _moment_is_measurements(moment):
+            return moment
+
+        return [
+            moment,
+            ops.Moment(self.qubit_noise_gate(q) for q in system_qubits)
+        ]
+
+
+class DepolarizingWithReadoutNoiseModel(devices.NoiseModel):
+    """DepolarizingNoiseModel with probabilistic bit flips preceding
+    measurement.
+
+    This simulates readout error.
+
+    If a circuit contains measurements, they be in moments that don't
+    also contain gates.
+
+    Args:
+        depol_prob: Depolarizing probability.
+        bitflip_prob: Probability of a bit-flip during measurement.
+    """
+
+    def __init__(self, depol_prob: float, bitflip_prob: float):
+        value.validate_probability(depol_prob, 'depol prob')
+        value.validate_probability(bitflip_prob, 'bitflip prob')
+        self.qubit_noise_gate = ops.DepolarizingChannel(depol_prob)
+        self.readout_noise_gate = ops.BitFlipChannel(bitflip_prob)
+
+    def noisy_moment(self, moment: 'cirq.Moment',
+                     system_qubits: Sequence['cirq.Qid']):
+        if _moment_is_measurements(moment):
+            return [
+                ops.Moment(self.readout_noise_gate(q) for q in system_qubits),
+                moment,
+            ]
+        return [
+            moment,
+            ops.Moment(self.qubit_noise_gate(q) for q in system_qubits),
```

Comment:
Why after instead of before like in the other one?

Reply:
I think you recently (read: very recently) changed it to noise models would keep measurements terminal. Some code expects terminal measurements.

Perhaps this is an unnecessary complication now that you're proposing putting all noise gates before